File: kotlin/src/main/kotlin/TennisGame1.kt
class TennisGame1(private val player1Name: String, private val player2Name: String) : TennisGame {

    private var player1Points: Int = 0
    private var player2Points: Int = 0

    override fun wonPoint(playerName: String) {
        if (playerName === "player1")
            player1Points += 1
        else
            player2Points += 1
    }

    override fun getScore(): String = when {
        isTie() -> getTieScore()
        isAdvantage() -> getAdvantageScore()
        else -> getMatchScore()
    }

    private fun isAdvantage() = player1Points >= 4 || player2Points >= 4

    private fun getAdvantageScore(): String {
        val minusResult = player1Points - player2Points
        return when {
            minusResult == 1 -> "Advantage player1"
            minusResult == -1 -> "Advantage player2"
            minusResult >= 2 -> "Win for player1"
            else -> "Win for player2"
        }
    }

    private fun getMatchScore(): String {
        var tempPoints: Int
        var score = ""
        for (index in 1..2) {
            val pair = determinePlayerForPoints(index, score)
            score = pair.first
            tempPoints = pair.second
            when (tempPoints) {
                0 -> score += "Love"
                1 -> score += "Fifteen"
                2 -> score += "Thirty"
                3 -> score += "Forty"
            }
        }
        return score
    }

    private fun determinePlayerForPoints(index: Int, score: String): Pair<String, Int> = when (index) {
        1 -> Pair(score, player1Points)
        else -> Pair("$score-", player2Points)
    }

    private fun isTie() = player1Points == player2Points

    private fun getTieScore(): String =
            when (player1Points) {
                0 -> "Love-All"
                1 -> "Fifteen-All"
                2 -> "Thirty-All"
                else -> "Deuce"
            }
}

File: swift/Tennis/TennisGame1.swift
import Foundation


class Player {
    let name: String
    var scorePoint: Int = 0

    init(name: String) {
        self.name = name
    }

    func increasePoints() {
        scorePoint+=1
    }
}

class Players {
    let player1: Player
    let player2: Player

    init(player1: Player,
         player2: Player) {
        self.player1 = player1
        self.player2 = player2
    }
    
    func calculateScorePoint(_ playerName: String) {
        getPlayer(playerName).increasePoints()
    }

    func getPlayer(_ playerName: String)-> Player{
        return (playerName == player1.name) ? player1 : player2
    }
}

class TennisGame1: TennisGame {
    private var players: Players

    required init(player1: String,
                  player2: String) {
        players = Players(player1: Player(name: player1),
                          player2: Player(name: player2))
    }

    var score: String? {
        if isTie() {
            return showTieMessage(currentScorePoint: players.player1.scorePoint)
        } else if isWinning() {
            return showWinningMessage()
        }
        return scoreLogic()
    }

    func wonPoint(_ playerName: String) {
        players.calculateScorePoint(playerName)
    }

    private func getLeadingPlayerName() -> String {
        return players.player1.scorePoint > players.player2.scorePoint ? players.player1.name : players.player2.name
    }

    private func showWinningMessage() -> String {
        return abs(scoreDifferential()) == 1 ? "Advantage \(getLeadingPlayerName())" : "Win for \(getLeadingPlayerName())"
    }

    private func scoreDifferential() -> Int {
        return players.player1.scorePoint-players.player2.scorePoint
    }

    fileprivate func scoreLogic() -> String {
        return scoreName(players.player1.scorePoint) + "-" + scoreName(players.player2.scorePoint)
    }

    fileprivate func showTieMessage(currentScorePoint: Int) -> String {
        return currentScorePoint>2
            ? "Deuce"
            : scoreName(currentScorePoint) + "-All"
    }

    

    fileprivate func isWinning() -> Bool {
        return players.player1.scorePoint>=4 || players.player2.scorePoint>=4
    }

    

    fileprivate func isTie() -> Bool {
        return players.player1.scorePoint == players.player2.scorePoint
    }

    fileprivate func scoreName(_ tempScore: Int) -> String {
        switch tempScore {
        case 0:
            return "Love"

        case 1:
            return "Fifteen"

        case 2:
            return "Thirty"

        case 3:
            return "Forty"

        default:
            return ""
        }
    }
}
